app/src/org/commcare/xml/AndroidBulkCaseXmlParser.java

```diff
@@ -0,0 +1,109 @@
+import org.commcare.engine.references.JavaHttpReference;
+import org.commcare.interfaces.HttpRequestEndpoints;
+import org.commcare.logging.AndroidLogger;
+import org.commcare.models.database.SqlStorage;
+import org.commcare.models.database.user.models.AndroidCaseIndexTable;
+import org.commcare.models.database.user.models.EntityStorageCache;
+import org.commcare.utils.FileUtil;
+import org.commcare.utils.GlobalConstants;
+import org.commcare.xml.bulk.BulkProcessingCaseXmlParser;
+import org.javarosa.core.io.StreamsUtil;
+import org.javarosa.core.reference.InvalidReferenceException;
+import org.javarosa.core.reference.Reference;
+import org.javarosa.core.reference.ReferenceManager;
+import org.javarosa.core.services.Logger;
+import org.javarosa.core.services.storage.IStorageUtilityIndexed;
+import org.javarosa.core.util.PropertyUtils;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * A bulk processing parser for the android platform. Provides superior performance when
+ * processing high case loads during syncing and/or processing.
+ *
+ * @author ctsims
+ */
+public class AndroidBulkCaseXmlParser extends BulkProcessingCaseXmlParser {
+    private HttpRequestEndpoints generator;
+    private final EntityStorageCache mEntityCache;
+    private final AndroidCaseIndexTable mCaseIndexTable;
+    private final SqlStorage<ACase> storage;
+
+    public AndroidBulkCaseXmlParser(KXmlParser parser,
+                                    SqlStorage<ACase> storage,
+                                    HttpRequestEndpoints generator) {
+        super(parser);
+        this.generator = generator;
+        mEntityCache = new EntityStorageCache("case");
+        mCaseIndexTable = new AndroidCaseIndexTable();
+        this.storage = storage;
+    }
+
+    protected SQLiteDatabase getDbHandle() {
+        return CommCareApplication.instance().getUserDbHandle();
+    }
+
+    @Override
+    protected Case buildCase(String name, String typeId) {
+        return new ACase(name, typeId);
+    }
+
+    @Override
+    protected void performBulkRead(Set<String> currentBulkReadSet, Map<String, Case> currentOperatingSet) throws InvalidStructureException, IOException, XmlPullParserException {
+        SQLiteDatabase db;
+        db = getDbHandle();
+        db.beginTransaction();
+        try {
+            for (ACase c : storage.getBulkRecordsForIndex(Case.INDEX_CASE_ID, currentBulkReadSet)) {
+                currentOperatingSet.put(c.getCaseId(), c);
+            }
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    @Override
+    protected void performBulkWrite(SortedMap<String, Case> writeLog) throws IOException {
```

Comment:
We want this list to only have one model in it per-write, but it's hard to guarantee that two objects which represent the same model actually will match with equality, so it seemed helpful to be super explicit about there needing to be a unique qualifier for it.

It's kind of awkward, but I think it will prevent errors long-term.